### PR TITLE
[Merged by Bors] - feat(RingTheory/Congruence): split `RingCon` construction and add some constructions

### PR DESCRIPTION
**Pull request by jjaassoonn**
added results are mainly about `RingCon` on the opposite ring

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Congruence/Basic.lean [hunk 1]
```diff
@@ -111,6 +111,18 @@ protected theorem mul {w x y z} : c w x → c y z → c (w * y) (x * z) :=
   c.mul'
 #align ring_con.mul RingCon.mul
 
+protected theorem sub {S : Type*} [AddGroup S] [Mul S] (t : RingCon S)
+    {a b c d : S} (h : t a b) (h' : t c d) : t (a - c) (b - d) := t.toAddCon.sub h h'
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Congruence/Basic.lean:115]: Does dot notation on `RingCon `not already work out of the box? by which I mean that without your new lemma I would expect `t.sub` to automatically expand to `t.toAddCon.sub`

File: Mathlib/RingTheory/Congruence/BigOperators.lean [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+# Interactions between `∑, ∏` and `RingCon`
+
+-/
+
+open BigOperators
+
+namespace RingCon
+
+/-- Congruence relation of a ring preserves finite sum indexed by a list. -/
+protected lemma list_sum {ι S : Type*} [AddCommMonoid S] [Mul S]
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Congruence/BigOperators.lean:19]:
```suggestion
protected lemma listSum {ι S : Type*} [AddCommMonoid S] [Mul S]
```
and below

File: Mathlib/RingTheory/Congruence/Basic.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/RingTheory/Congruence/Basic.lean:115]: Ah maybe it proves `t.toAddCon (a - c) (b - d)`... Mind checking?

Comment (jjaassoonn) [reply Mathlib/RingTheory/Congruence/Basic.lean:115]: The following gave an error
```lean
example  {S : Type*} [AddGroup S] [Mul S] (t : RingCon S) : true := by
  have := t.sub

/-
invalid field 'sub', the environment does not contain 'RingCon.sub'
  t
has type
  RingCon S
-/
```

File: Mathlib/RingTheory/Congruence/BigOperators.lean (hunk 2 above)
Comment (jjaassoonn) [reply Mathlib/RingTheory/Congruence/BigOperators.lean]: thanks

File: Mathlib/RingTheory/Congruence/Opposite.lean [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+/--
+If `c` is a `RingCon R`, then `(a, b) ↦ c b.unop a.unop` is a `RingCon Rᵐᵒᵖ`.
+-/
+def op (c : RingCon R) : RingCon Rᵐᵒᵖ where
+  __ := c.toCon.op
+  mul' h1 h2 := c.toCon.op.mul h1 h2
+  add' h1 h2 := c.add h1 h2
+
+lemma op_iff {c : RingCon R} {x y : Rᵐᵒᵖ} : c.op x y ↔ c (y.unop) (x.unop) := Iff.rfl
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Congruence/Opposite.lean:30]:
```suggestion
lemma op_iff {c : RingCon R} {x y : Rᵐᵒᵖ} : c.op x y ↔ c y.unop x.unop := Iff.rfl
```

File: Mathlib/RingTheory/Congruence/Opposite.lean [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+-/
+def unop (c : RingCon Rᵐᵒᵖ) : RingCon R where
+  __ := c.toCon.unop
+  mul' h1 h2 := c.toCon.unop.mul h1 h2
+  add' h1 h2 := c.add h1 h2
+
+lemma unop_iff {c : RingCon Rᵐᵒᵖ} {x y : R} : c.unop x y ↔ c (.op y) (.op x) := Iff.rfl
+
+/--
+The congruences of a ring `R` bijects to the congrunences of the opposite ring `Rᵐᵒᵖ`.
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Congruence/Opposite.lean:43]:
```suggestion
The congruences of a ring `R` biject to the congruences of the opposite ring `Rᵐᵒᵖ`.
```

File: Mathlib/RingTheory/Congruence/Opposite.lean [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+-/
+@[simps]
+def orderIsoOp : RingCon R ≃o RingCon Rᵐᵒᵖ where
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Congruence/Opposite.lean:46]:
```suggestion
def opOrderIso : RingCon R ≃o RingCon Rᵐᵒᵖ where
```

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer merge

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/GroupTheory/Congruence/BigOperators.lean [hunk 6]
```diff
@@ -20,7 +20,7 @@ namespace Con
 
 /-- Multiplicative congruence relations preserve product indexed by a list. -/
 @[to_additive "Additive congruence relations preserve sum indexed by a list."]
-protected theorem list_prod {ι M : Type*} [CommMonoid M] (c : Con M) {l : List ι} {f g : ι → M}
+protected theorem list_prod {ι M : Type*} [Monoid M] (c : Con M) {l : List ι} {f g : ι → M}
```
[the remark this answers is not in the document]
Comment (rosborn) [reply Mathlib/GroupTheory/Congruence/BigOperators.lean:23]: Gonna copy my comment from another PR which I've closed.
"Technically, [this does not need associativity]. If we really wanted to generalize, we should have api for a product over a binary tree (or free magma). I do have a feeling that mathlib is using lists and multisets for things that are morally free monoids and free commutative monoids, but that might be a refactor for another day."

[the remark this answers is not in the document]
Comment (jjaassoonn) [reply Mathlib/GroupTheory/Congruence/BigOperators.lean]: `List.prod_cons` requires `Monoid M`. I think removing associativity here means we need to remove associativity at `List.prod_cons` as well.

[the remark this answers is not in the document]
Comment (rosborn) [reply Mathlib/GroupTheory/Congruence/BigOperators.lean:23]: As `prod` is defined with `foldl`, `prod_cons` requires associativity. You would need to unfold `prod` and work with `foldl` directly to prove this without associativity. I would argue this generalization requires more work and is outside of the scope of this PR.

File: Mathlib/RingTheory/Congruence/Opposite.lean [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+the opposite ring `Rᵐᵒᵖ`.
+
+-/
+
+variable {R : Type*} [Add R] [Mul R]
+
+namespace RingCon
+
+/--
+If `c` is a `RingCon R`, then `(a, b) ↦ c b.unop a.unop` is a `RingCon Rᵐᵒᵖ`.
```
Comment (eric-wieser) [inline Mathlib/RingTheory/Congruence/Opposite.lean:23]: Any idea why we swap the argument order?

Comment (eric-wieser) [reply Mathlib/RingTheory/Congruence/Opposite.lean:23]: Ah, I guess this question is better directed at https://github.com/leanprover-community/mathlib4/pull/13610/files#r2032230268

**Pull request closed without merging**